### PR TITLE
Multi-Class Logistic Regression Training and Evaluation

### PR DESCRIPTION
**Pull request by axiomcura**
This PR  implements the training of the Mulit-Class Logistic Regression model.

The multi-class models have undergone training and evaluation under various datasets, including:

- Testing and Training datasets
- Utilization of both shuffled and unshuffled datasets (features space shuffled).
- Implementation of holdouts, including plate, treatment, and well differentiations.

Evaluation metrics were derived from F1 scores, precision, and recall.

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -80,3 +81,45 @@ def drop_na_samples(
     ).reset_index(drop=True)
 
     return profile
+
+
+def shuffle_features(feature_vals: np.array, seed: Optional[int] = 0) -> np.array:
+    """suffles all values within feature space
```
Comment (d33bs) [inline src/utils.py]:
```suggestion
    """Shuffles all values within feature space
```

File: src/utils.py [hunk 2]
```diff
@@ -80,3 +81,45 @@ def drop_na_samples(
...
+
+    Parameters
+    ----------
+    feature_vals : np.array
+        shuffled
```
Comment (d33bs) [inline src/utils.py]:
```suggestion
        Values to be shuffled.
```

File: src/utils.py [hunk 3]
```diff
@@ -80,3 +81,45 @@ def drop_na_samples(
...
+    # setting seed
+    np.random.seed(seed)
+
+    # shuffle given array
+    if not isinstance(feature_vals, np.ndarray):
+        raise TypeError("'feature_vals' must be a numpy array")
+    if feature_vals.ndim != 2:
+        raise TypeError("'feature_vals' must be a 2x2 matrix")
+
+    # creating a copy for feature vales to prevent overwriting of global variables
```
Comment (d33bs) [inline src/utils.py]:
```suggestion
    # creating a copy for feature values to prevent overwriting of global variables
```

File: src/utils.py [hunk 4]
```diff
@@ -80,3 +81,45 @@ def drop_na_samples(
...
+    feature_vals = np.copy(feature_vals)
```
Comment (d33bs) [inline src/utils.py]: Double checking, would it make sense to remove this copy due to `np.random.permutation` implicitly providing a new copy of the data ([link](https://stackoverflow.com/a/15474335))?

File: notebooks/0.freature_selection/feature_processing.ipynb
Comment (d33bs) [inline notebooks/0.freature_selection/feature_processing.ipynb:1]: Would this file need a related `.py` file update?

File: notebooks/1.data_splits/data_splits.ipynb
Comment (d33bs) [inline notebooks/1.data_splits/data_splits.ipynb:1]: Would this file need a related `.py` file update?

File: notebooks/2.modeling/nbconvert/modeling.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+#
+
+# In[7]:
+
+
+# train and get the best_model
+best_model = train_multiclass(X_train, y_train, param_grid=param_grid, seed=0)
+
+
+# In[ ]:
```
Comment (d33bs) [inline notebooks/2.modeling/nbconvert/modeling.py]: Double checking, does this cell (and the others below) need to be re-run before merging?

File: notebooks/2.modeling/nbconvert/modeling.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+    "estimator__l1_ratio": uniform(0, 1),
+}
+
+
+# In[4]:
+
+
+# loading injury codes
+with open(data_splits_dir / "injury_codes.json") as json_file:
+    injury_codes = json.load(json_file)
```
Comment (d33bs) [inline notebooks/2.modeling/nbconvert/modeling.py]: Would it make sense to use [`pd.read_json()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html) here instead of the call to `open`?

File: results/1.data_splits/injury_codes.json
Comment (d33bs) [inline results/1.data_splits/injury_codes.json:1]: This was cool to see, I thought the related code you added related to this data was clever. Seeing how the values duplicate I wondered if you could leverage something like a [`pd.DataFrame.transpose()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.transpose.html) (doesn't need to be Pandas, just an example) to help store the values once and transform the data when needed in various formats. 

For example:

```python
import pandas as pd

df = pd.DataFrame({'col1': [1, 2], 'col2': [3, 4]})

print(df.head(), end="\n\n")

print(df.T.head())

#    col1  col2
# 0     1     3
# 1     2     4

#       0  1
# col1  1  2
# col2  3  4
```

**Review by d33bs — APPROVED** [review]
Nice job! I left a few comments and suggestions. Please don't hesitate to let me know if you have any questions.

File: notebooks/0.freature_selection/feature_processing.ipynb [hunk 7]
```diff
@@ -600,7 +600,7 @@
                 "\n",
                 "# drop wells that do not have an injury\n",
                 "injured_df = injured_df.dropna(subset=\"injury_type\").reset_index(drop=True)\n",
-                "print(\"Number of wells\", len(injured_df[\"Plate\"].unique()))\n",
+                "print(\"Number of plates\", len(injured_df[\"Plate\"].unique()))\n",
```
Comment (gwaybio) [inline notebooks/0.freature_selection/feature_processing.ipynb:603]: See the name of the folder - please try to catch these typos (a close attention to detail will pay off in the long run)

File: notebooks/1.data_splits/nbconverted/data_splits.py [hunk 8]
```diff
@@ -247,25 +268,84 @@
...
+for injury_type, df in fs_profile_df.groupby("injury_type"):
+    for treatment, df2 in df.groupby("Compound Name"):
+        n_wells = df2.shape[0]
+        injury_treatment_metadata.append([injury_type, treatment, n_wells])
+
+# convert to df
+injury_treatment_metadata = pd.DataFrame(
+    injury_treatment_metadata, columns="injury_type treatment n_wells".split()
+)
+injury_treatment_metadata
```
Comment (gwaybio) [inline notebooks/1.data_splits/nbconverted/data_splits.py:293]: Ok for now (it does the job), but I think pandas can do all of this for you. Maybe something like:

```pandas
injury_treatment_metadata = (
    fs_profile_df
    .groupby(["injury_type", "Compound Name"])
    .value_counts()
    .reset_index()
)
```

Would have the same effect and be more readable?

File: notebooks/1.data_splits/nbconverted/data_splits.py [hunk 9]
```diff
@@ -247,25 +268,84 @@
...
+
+# In[13]:
+
+
+# setting random seed
+seed = 0
+min_treatments_per_injury = 10
+
+# setting random seed globally
+np.random.seed(seed)
```
Comment (gwaybio) [inline notebooks/1.data_splits/nbconverted/data_splits.py:304]: Typically, constants and random seeds are set in the beginning of notebooks, after imports and function definitions (when you establish file names and paths)

File: notebooks/1.data_splits/nbconverted/data_splits.py [hunk 10]
```diff
@@ -247,25 +268,84 @@
...
+print("Below are the accepted cell injuries and treatments to be held out")
+selected_treatments_to_holdout
+
+
+# In[14]:
 
 
-#### Plate heldout dataset
 seed = 0
 n_samples = 15
```
Comment (gwaybio) [inline notebooks/1.data_splits/nbconverted/data_splits.py:338]: Do you still need these constants?

File: notebooks/2.modeling/nbconverted/modeling.py [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+
+# In[2]:
+
+
+def train_multiclass(
+    X_train: np.ndarray, y_train: np.ndarray, param_grid: dict, seed: Optional[int] = 0
+) -> BaseEstimator:
+    """Develops a Logistic Regression model employing the One vs Rest training
+    scheme and incorporates RandomizedSearchCV for optimal parameter selection.
+    This approach utilizes RandomizedSearchCV to explore a range of parameters
```
Comment (gwaybio) [inline notebooks/2.modeling/nbconverted/modeling.py]: consider removing this sentence as it is redundant with previous sentence

File: notebooks/2.modeling/nbconverted/modeling.py [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+
+# then we can split the data set with are newly binarized labels
+# we made sure to use stratify to ensure proportionality within training and test datasets
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=seed, stratify=y)
+
+
+# In[6]:
+
+
+def evaluate(
```
Comment (gwaybio) [inline notebooks/2.modeling/nbconverted/modeling.py]: functions are typically defined after imports (prior to any data wrangling)

File: notebooks/2.modeling/nbconverted/modeling.py [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+        precision, recall, _ = precision_recall_curve(y[:, i], probability[:, i])
+
+        # iterate all scores and save all data into a list
+        for i in range(len(precision)):
+            precision_recall_scores.append([dataset, shuffled, precision[i], recall[i]])
+
+    # creating scores df
+    precision_recall_scores = pd.DataFrame(
+        precision_recall_scores, columns=["dataset", "shuffled", "precision", "recall"]
+    )
```
Comment (gwaybio) [inline notebooks/2.modeling/nbconverted/modeling.py]: This feels like it can be easily optimized (although, if it works, it works!) Up to you.

Something like this might work:

```python
precision_recall_scores = pd.DataFrame([precision, recall]).transpose()
precision_recall_scores = (
    precision_recall_scores
    .assign(dataset=dataset, shuffled=shuffled)
)
```

(I didn't test any of these suggestions)

File: notebooks/2.modeling/nbconverted/modeling.py [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+
+# In[ ]:
+
+
+shuffle_test_precision_recall_df, shuffle_test_f1_score_df = evaluate(
+    model=best_model, X=X_test, y=y_test, dataset="test", shuffled=True, seed=0
+)
+shuffle_train_precision_recall_df, shuffle_train_f1_score_df = evaluate(
+    model=best_model, X=X_train, y=y_train, dataset="train", shuffled=True, seed=0
+)
```
Comment (gwaybio) [inline notebooks/2.modeling/nbconverted/modeling.py:137]:
```suggestion
shuffle_train_precision_recall_df, shuffle_train_f1_score_df = evaluate(
    model=best_model, X=shuffled_X_train, y=y_train, dataset="train", shuffled=True, seed=0
)
```

You need to evaluate on the data you trained on - a well controlled model has high training set performance, but low (close to random) performance in the test sest.

File: notebooks/2.modeling/nbconverted/modeling.py [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+
+
+# evaluating with plate holdout
+plate_ho_precision_recall_df, plate_ho_f1_score_df = evaluate(
+    model=best_model,
+    X=X_plate_holdout,
+    y=y_plate_holout,
+    dataset="plate_holdout",
+    shuffled=False,
+    seed=0,
```
Comment (gwaybio) [inline notebooks/2.modeling/nbconverted/modeling.py:186]: you set the `seed` variable above to the same thing, so consider using it here

Comment (gwaybio) [reply notebooks/2.modeling/nbconverted/modeling.py:186]: (also below)

File: notebooks/2.modeling/nbconverted/modeling.py [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+#
+
+# In[7]:
+
+
+# train and get the best_model
+best_model = train_multiclass(X_train, y_train, param_grid=param_grid, seed=0)
+
+
+# In[ ]:
```
Comment (gwaybio) [inline notebooks/2.modeling/nbconverted/modeling.py]: It looks like this notebook wasn't fully executed (it should be added after sequential and complete execution)

File: src/utils.py [hunk 17]
```diff
@@ -4,6 +4,7 @@
 
 from typing import Optional
 
+import numpy as np
```
Comment (gwaybio) [inline src/utils.py:10]: You could consider adding the other functions you've defined in this `utils.py` file here as well (e.g., evaluate, train, etc.)

**Review by gwaybio — APPROVED** [review]
Well done Erik! I love the modularity in this, and it is clear you're thinking like a data scientist.

Three general comments.
1) The biggest thing I noticed is that you should be evaluating the shuffled model with the shuffled x data
2) I made a couple optional optimization comments
3) Other minor comments

I don't need to perform a re-review, once you make the necessary changes (only comment 1 above is necessary)

Thanks!

File: src/utils.py (hunk 4 above)
Comment (axiomcura) [reply src/utils.py]: Correct! Thank you! Removed!

**Comment by axiomcura** [comment]
Thank you all! I have attended all suggestions!
